modules/polygon/src/polygon-utilities.d.ts

```diff
@@ -0,0 +1,80 @@
+
+type SegmentVisitor = (p1: number[], p2: number[], i1: number, i2: number) => void;
+
+type SegmentVisitorFlat = (p1x: number, p1y: number, p2x: number, p2y: number, i1: number, i2: number) => void;
+
+export const WINDING_CLOCKWISE: number;
+export const WINDING_COUNTER_CLOCKWISE: number;
+
+/**
+ * Checks winding direction of the polygon and reverses the polygon in case if opposite winding direction.
+ * Note: points are modified in-place.
+ * @param points Array of points.
+ * @param direction Requested winding direction. 1 is for clockwise, -1 for counterclockwise winding direction.
+ */
+export function ensurePolygonWindingDirection(points: number[][], direction: number): void;
+
+/**
+ * Returns winding direction of the polygon.
+ * @param points Array of points.
+ * @param direction Requested winding direction.
+ * @returns Winding direction of the polygon.
+ */
+export function getPolygonWindingDirection(points: number[][]): number;
+
+/**
+ * Returns signed area of the polygon.
+ * @param points Array of points.
+ * @returns Signed area of the polygon.
+ */
+export function getPolygonSignedArea(points: number[][]): number;
+
+/**
+ * Calls visitor callback for each segment in the polygon.
+ * @param points Array of points.
+ * @param visitor A callback to call for each segment.
+ */
+export function forEachSegmentInPolygon(points: number[][], visitor: SegmentVisitor): void;
+
+/**
+ * Checks winding direction of the polygon and reverses the polygon in case if opposite winding direction.
+ * Note: points are modified in-place.
+ * @param points An array that represents points of the polygon.
+ * @param start Start index of the polygon in the points array.
+ * @param end End index of the polygon in the points array.
+ * @param size Size of a point, 2 (XZ) or 3 (XYZ).
+ * @param direction Requested winding direction. 1 is for clockwise, -1 for counterclockwise winding direction.
+ */
+export function ensurePolygonWindingDirectionFlat(points: number[], start: number, end: number, size: number, direction: number): void;
+
+/**
+ * Returns winding direction of the polygon.
+ * @param points An array that represents points of the polygon.
+ * @param start Start index of the polygon in the points array.
+ * @param end End index of the polygon in the points array.
+ * @param size Size of a point, 2 (XZ) or 3 (XYZ).
+ * @returns Winding direction of the polygon.
+ */
+export function getPolygonWindingDirectionFlat(points: number[], start: number, end: number, size: number): number;
```

Comment:
Should we create an `options?: object` parameter for all of these flat function?
- so we can have `start` and `end `parameters default to the entire array?
- And `size` could default to something, maybe `3`?

Reply:
These utilities are called many times when we iterate through input data. I'd prefer not having to construct small parameter objects.

Reply:
@ibgreen @Pessimistress Let me know what we want here. 
- we want skipIsClosedCheck somehow passed to forEachSegment(), but we are out of # of arguments per current lint settings.
- at the same time we don't want to create options objects on each level

Reply:
Linter rules can be overridden for a specific function, if appropriate. This library should be performance first. 

Please work to address the suggestions. However since you already have a benchmark infrastructure set up, you could also do some tests and make sure that all these suggested optimizations make a meaningful difference. 
- Not creating a function may or may not be faster than creating a single capturing function - it is hard to predict JavaScript JIT performance.
- Comparing first and last coordinate may take say less than 1% of function run time and may or may not be worth a parameter to avoid the check etc .

Make sure you run benchmarks in the browser `yarn test bench-browser`, my experience is that recent versions of Node.js usually gets better results (maybe has a more modern JS engine).